src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -1216,4 +1250,31 @@ public long countSubmissions(Long exerciseId) {
     public long countSubmissionsToAssessByCourseId(Long courseId) {
         return programmingExerciseRepository.countByCourseIdSubmittedBeforeDueDate(courseId);
     }
+
+    private Map<Path, ScheduledFuture> futures = new HashMap<>();
```

Comment:
Why do we need to store the futures in this map?

Reply:
We could also move this functionality to an inner class service. That way, we still keep the member variables close to the relevant methods while still having all variable declarations at the top of the class.

Reply:
I don't know, good point for a subsequent PR